### DISCUSSION TITLE
Variable Length (VarLen) Kernel

### DISCUSSION THREAD

**Discussion by fountaindive** [google/flax#5587] [Q&A]
Hi everyone,

I'm looking to use variable-length (unpadded/ragged batching) attention kernels in JAX/Flax to avoid padding overhead.

A few questions:

Does JAX or Flax directly support variable-length attention kernels (e.g., FlashAttention varlen)?

Is this handled via jax.nn.dot_product_attention using implementation='cudnn' along with query_seq_lengths and key_value_seq_lengths?

If anyone has a minimal working example demonstrating the expected input shapes (e.g., whether sequences should be flattened into a single batch dimension or passed with dynamic length vectors), I'd really appreciate it!

Thanks!

**Comment by dfbustosus** [comment]
Yes. Current JAX supports sequence-length-aware attention, with one important distinction: the public API uses **padded dense tensors plus per-example length vectors**, not FlashAttention-style flattened packed tokens with cumulative offsets.

The batched shapes are:

- Query: `[B, T, N, H]`
- Key/value: `[B, S, K, H]`
- Query lengths: `[B]`, `int32`
- Key/value lengths: `[B]`, `int32`
- Output: `[B, T, N, H]`

`N` is the number of query heads and `K` is the number of key/value heads. `N` must be divisible by `K`; `K == N` is MHA, `K == 1` is MQA, and intermediate divisors are GQA. Each sequence's valid tokens occupy the prefix of its padded row.

```python
import jax
import jax.numpy as jnp


def varlen_attention(q, k, v, q_lengths, kv_lengths, *, causal=False):
    return jax.nn.dot_product_attention(
        q,
        k,
        v,
        is_causal=causal,
        query_seq_lengths=jnp.asarray(q_lengths, dtype=jnp.int32),
        key_value_seq_lengths=jnp.asarray(kv_lengths, dtype=jnp.int32),
        implementation="cudnn",
    )


B, T, S, N, K, H = 2, 128, 160, 8, 8, 64
q = jnp.zeros((B, T, N, H), dtype=jnp.bfloat16)
k = jnp.zeros((B, S, K, H), dtype=jnp.bfloat16)
v = jnp.zeros((B, S, K, H), dtype=jnp.bfloat16)

q_lengths = jnp.array([93, 61], dtype=jnp.int32)
kv_lengths = jnp.array([145, 72], dtype=jnp.int32)

y = varlen_attention(q, k, v, q_lengths, kv_lengths)
assert y.shape == (B, T, N, H)
```

Do not concatenate the sequences into `[total_tokens, ...]`; `jax.nn.dot_product_attention` does not expose a public packed-offset interface.

A few limitations matter:

- `implementation=None` currently selects the XLA implementation. Specify `"cudnn"` to request fused cuDNN attention.
- XLA applies the lengths semantically but still builds dense masked logits, so it does not remove padded computation.
- cuDNN is CUDA-only and accepts selected shapes and dtypes. The ordinary path uses `float16` or `bfloat16`; unsupported combinations raise instead of silently falling back.
- Arrays are still allocated at the batch maximum `T` and `S`. This is not the same API as packed FlashAttention varlen.
- `jax.nn.dot_product_attention` does not expose attention-probability dropout.

Current Flax Linen and NNX attention wrappers expose masks but not `query_seq_lengths` or `key_value_seq_lengths`. Linen also constructs a fixed set of arguments for its `attention_fn`, so extra length arguments cannot simply be passed through `MultiHeadDotProductAttention.__call__`.

The reliable Flax approach is therefore a small custom Linen/NNX block: use the normal projections to create `[B,T,N,H]` and `[B,S,K,H]`, call the helper above, and apply the output projection. I would not depend on private `jax._src.cudnn` packed-offset APIs.

References: [JAX public API and validation](https://github.com/jax-ml/jax/blob/jax-v0.11.2/jax/_src/nn/functions.py#L1086-L1286), [length/cuDNN tests](https://github.com/jax-ml/jax/blob/jax-v0.11.2/tests/nn_test.py#L282-L337), [cuDNN restrictions](https://github.com/jax-ml/jax/blob/jax-v0.11.2/jax/_src/cudnn/fused_attention_stablehlo.py#L275-L400), [Flax Linen wrapper](https://github.com/google/flax/blob/v0.12.10/flax/linen/attention.py#L436-L479), and [Flax NNX wrapper](https://github.com/google/flax/blob/v0.12.10/flax/nnx/nn/attention.py#L737-L752).
